File: math/euler/ami.py
#!/usr/bin/env python
from pprint import pprint
import math
import sys
import time

def even_div(num,div):
	return float(num)/div == num/div

def find_even_div(num):
	even_divs = []
	for div in range(1,int(math.sqrt(num))+1):
		if even_div(num,div):
			even_divs.append(div)
			even_divs.append(num/div)
	even_divs = remove_dup(even_divs)
	even_divs.remove(num)
	return even_divs

def remove_dup(l):
	ret = []
	for i in l:
		if not i in ret:
			ret.append(i)
	return ret

def find_sum(l):
	s = 0
	for i in l:
		s += i
	return s

def find_amis(total):
	sums = [0]
	amis = []
	for i in range(1,total+1):
		s = find_sum(find_even_div(i))
		sums.append(s)
	for i in range(1,total+1):
		try:
			if sums[sums[i]] == i and i != sums[i]:
				amis.append(i)
		except IndexError:
			pass
	return find_sum(amis)

if len(sys.argv) == 2:
	start = time.time()
	print find_amis(int(sys.argv[1]))
	print 'time:'+str(time.time()-start)

File: math/euler/ami.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <time.h>

int even_div(int num, int div){
	return fmod((float)num,(float)div) == 0.0;
}

int can_add(int num, int* array, int size){
	int i;
	int can = 1;
	for(i=0;i<size;i++){
		if(num == array[i]){
			can = 0;
		}
	}
	return can;
}

int find_sum(int* array, int size){
	int total = 0;
	int i;
	for(i=0;i<size;i++){
		total = total + array[i];
	}
	return total;
}


int find_even_divs(int num, int* even_divs){
	int div;
	int i;
	int current_index = 0;
	even_divs[current_index] = 1;
	current_index++;
	for(div=2;div<num;div++){
		if(even_div(num,div)){
			if(can_add((int)(num/div),even_divs,current_index)){
				even_divs[current_index] = (int)(num/div);
				current_index++;
			}
			if(can_add(div,even_divs,current_index)){
				even_divs[current_index] = div;
				current_index++;
			}
		}
	}
	return current_index;
}

int find_amis(int top){
	int sums[top];
	int i;
	int s;
	int count;
	int current_index = 0;
	int amis[top];
	int even_divs[100];
	for(i=0;i<top;i++){
		count = find_even_divs(i,even_divs);
		s = find_sum(even_divs,count);
		sums[i] = s;
	}
	for(i=0;i<top;i++){
		if(sums[i] < top && sums[i] > 0){
			if(sums[sums[i]] == i && i != sums[i]){
				amis[current_index] = i;
				current_index++;
			}
		}
	}
	return find_sum(amis,current_index);
}

int main(int argc, char** argv){
	int to_find = 10000;
	if(argc==2){
		to_find = atoi(argv[1]);
	}
	int start_time = time(NULL);
	int sum = find_amis(to_find);
	printf("%d\n",sum);
	printf("%d seconds\n",time(NULL)-start_time);
	return 0;
}
